src/Lykke.Service.BitfinexAdapter.Services/Exchange/BitfinexExchange.cs

```diff
@@ -0,0 +1,238 @@
+﻿using Common.Log;
+using Lykke.Service.BitfinexAdapter.Core.Domain;
+using Lykke.Service.BitfinexAdapter.Core.Domain.Exceptions;
+using Lykke.Service.BitfinexAdapter.Core.Domain.RestClient;
+using Lykke.Service.BitfinexAdapter.Core.Domain.Settings;
+using Lykke.Service.BitfinexAdapter.Core.Domain.Trading;
+using Lykke.Service.BitfinexAdapter.Core.Domain.Trading.Enums;
+using Lykke.Service.BitfinexAdapter.Core.RestClient;
+using Lykke.Service.BitfinexAdapter.Core.Utils;
+using Lykke.Service.BitfinexAdapter.Models;
+using Lykke.Service.BitfinexAdapter.Services.OrderBooksHarvester;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+//using Position = Lykke.Service.BitfinexAdapter.Core.Domain.Trading.Position;
+
+namespace Lykke.Service.BitfinexAdapter.Services.Exchange
+{
+    public class BitfinexExchange : ExchangeBase
+    {
+        private readonly BitfinexModelConverter _modelConverter;
+        private readonly BitfinexOrderBooksHarvester _orderBooksHarvester;
+        //private readonly BitfinexExecutionHarvester _executionHarvester;
+        private readonly IBitfinexApi _exchangeApi;
+
+        public BitfinexExchange(BitfinexAdapterSettings configuration,
+            //TranslatedSignalsRepository translatedSignalsRepository,
+            BitfinexOrderBooksHarvester orderBooksHarvester,
+            //BitfinexExecutionHarvester executionHarvester, 
+            ILog log)
+            : base(Constants.BitfinexExchangeName, configuration, log)
+        {
+            _modelConverter = new BitfinexModelConverter(configuration);
+            _orderBooksHarvester = orderBooksHarvester;
+            //_executionHarvester = executionHarvester;
+            var credenitals = new BitfinexServiceClientCredentials(configuration.ApiKey, configuration.ApiSecret);
+            _exchangeApi = new BitfinexApi(credenitals)
+            {
+                BaseUri = new Uri(configuration.EndpointUrl)
+            };
+
+
+
+            orderBooksHarvester.MaxOrderBookRate = configuration.MaxOrderBookRate;
+        }
+
+        public override async Task<ExecutionReport> AddOrderAndWaitExecution(TradingSignal signal, TimeSpan timeout)
+        {
+            var symbol = _modelConverter.LykkeSymbolToExchangeSymbol(signal.Instrument.Name);
+            var volume = signal.Volume;
+            var orderType = _modelConverter.ConvertOrderType(signal.OrderType);
+            var side = _modelConverter.ConvertTradeType(signal.TradeType);
+            var price = signal.Price == 0 ? 1 : signal.Price ?? 1;
+            var cts = new CancellationTokenSource(timeout);
+
+            var response = await _exchangeApi.AddOrder(symbol, volume, price, side, orderType, cts.Token);
+
+            if (response is Error error)
```

Comment:
Pattern matching using type checking is not a good idea. Throwing exceptions, introducing `Result<T>` type, return `(result?, error?)` tuple is better. Compile-time checking, exposing intention etc

Reply:
The same for other methods

Reply:
This has been refactored and will be reviewed in the next PR (its in a different branch)